packages/inventory/src/services/reservation-item.ts

```diff
@@ -0,0 +1,271 @@
+        itemRepository.save(inventoryItem),
+        this.inventoryLevelService_
+          .withTransaction(manager)
+          .adjustReservedQuantity(
+            data.inventory_item_id,
+            data.location_id,
+            data.quantity
+          ),
+      ])
+
+      return newInventoryItem
+    })
+
+    await this.eventBusService_.emit(ReservationItemService.Events.CREATED, {
+      id: result.id,
+    })
+
+    return result
+  }
+
+  /**
+   * Update a reservation item.
+   * @param {string} reservationItemId - The reservation item's id.
+   * @param {DeepPartial<ReservationItem>} data - The reservation item data to update.
+   * @returns {Promise<ReservationItem>} - The updated reservation item.
+   */
+  async update(
+    reservationItemId: string,
+    data: Omit<
+      DeepPartial<ReservationItem>,
+      "id" | "created_at" | "metadata" | "deleted_at"
+    >
+  ): Promise<ReservationItem> {
+    const item = await this.atomicPhase_(async (manager) => {
+      const itemRepository = manager.getRepository(ReservationItem)
+
+      const item = await this.retrieve(reservationItemId)
+
+      const shouldUpdateQuantity =
+        isDefined(data.quantity) && data.quantity !== item.quantity
+
+      const ops: Promise<unknown>[] = []
+      if (shouldUpdateQuantity) {
+        const quantityDiff = data.quantity! - item.quantity
+        ops.push(
+          this.inventoryLevelService_
+            .withTransaction(manager)
+            .adjustReservedQuantity(
+              item.inventory_item_id,
+              item.location_id,
+              quantityDiff
+            )
+        )
+      }
+
+      itemRepository.merge(item, data)
+      ops.push(itemRepository.save(item))
+
+      await Promise.all(ops)
+
+      return item
+    })
+
+    await this.eventBusService_.emit(ReservationItemService.Events.UPDATED, {
+      id: item.id,
+    })
+
+    return item
+  }
+
+  /**
+   * Deletes a reservation item by line item id.
+   * @param {string} lineItemId - the id of the line item to delete.
+   * @returns {Promise<void>} - an empty promise
+   */
+  async deleteByLineItem(lineItemId: string): Promise<void> {
+    await this.atomicPhase_(async (manager) => {
+      const itemRepository = manager.getRepository(ReservationItem)
+
+      const items = await this.list({ line_item_id: lineItemId })
+
+      const ops: Promise<unknown>[] = []
+      for (const item of items) {
+        ops.push(itemRepository.softRemove({ line_item_id: lineItemId }))
+        ops.push(
+          this.inventoryLevelService_
+            .withTransaction(manager)
+            .adjustReservedQuantity(
+              item.inventory_item_id,
+              item.location_id,
+              item.quantity * -1
+            )
+        )
+      }
+      await Promise.all(ops)
+    })
+
+    await this.eventBusService_.emit(
+      ReservationItemService.Events.DELETED_BY_LINE_ITEM,
+      {
+        line_item_id: lineItemId,
+      }
+    )
+  }
```

Comment:
thought (non-blocking): wonder if the existence of a `line_item_id` is a bad pattern as it couples the InventoryService with the design of the core. 

It could be argued that a cleaner approach would be to have a `LineItemReservations` table in the core with:
- `line_item_id`: id of a cart/order line item
- `reservation_item_id`: id of a reservation in the inventory system

It is not a big deal for me, but would like to hear your opinion. (cc: @pKorsholm)

Reply:
In this case to me makes more sense the inventory service managing reservations and holding the external relation (line_item_id).
Having a LineItemReservation would transfer the responsibility of managing reservations to the line item, which is somewhat duplicating the business logic.